Introduce sanity/compatibility test for live clusters

ryoqun opened the pull request:
#### Problem

**bisecting is hurting...** (hard-labored fruit this time: https://github.com/solana-labs/solana/pull/12176)

#### Summary of Changes

I think if ci time and resource is allowed, this should be run on each prs instead of nightly ci job. and it seems that running this doesn't take much time.

~~- [ ] todo what to do if the tested cluster is dead? Maybe easy turn-off knob like github's `skip-live-cluster` label?~~ (EDIT: Well, let's skip this for now? clusters are pretty stable nowadays)

~~- [ ] todo if the cluster is dead, fallback to some periodic backup of snapshot + minimum ledger?~~ (EDIT: Well, let's skip this for now? clusters are pretty stable nowadays)

ryoqun commented on runtime/src/bank.rs:
```diff
@@ -110,7 +110,7 @@ type TransactionAccountRefCells = Vec<Rc<RefCell<Account>>>;
 
 // Eager rent collection repeats in cyclic manner.
-// Each cycle is composed of <partiion_count> number of tiny pubkey subranges
+// Each cycle is composed of <partition_count> number of tiny pubkey subranges
```
let's increase my karma. :)

ryoqun commented on ci/buildkite-pipeline.sh:
```diff
@@ -125,8 +125,9 @@ wait_step() {
 all_test_steps() {
-  command_step checks ". ci/rust-version.sh; ci/docker-run.sh \$\$rust_nightly_docker_image ci/test-checks.sh" 20
-  wait_step
+  #command_step checks ". ci/rust-version.sh; ci/docker-run.sh \$\$rust_nightly_docker_image ci/test-checks.sh" 20
```
Obviously, I need to revert these before merging!

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,149 @@
+    # for your pain-less copy-paste
+
+# UPDATE docs/src/clusters.md TOO!!
+test_with_live_cluster "testnet" \
```
When backporting to v1.2, I'll remove this line.

ryoqun commented on docs/src/clusters.md:
```diff
@@ -74,20 +74,21 @@ solana config set --url https://testnet.solana.com
 
 ##### Example `solana-validator` command-line
 
+[comment]: <> (UPDATE ci/live-cluster-sanity.sh TOO!)
```
TIL: https://stackoverflow.com/a/20885980

ryoqun commented:
@mvines I think this pr is getting in pretty good shape. Could you review this? I changed to launch and run on adhoc GCE instance and the test duration is pretty short (~10 min) for both testnet and mainnet-beta.

ryoqun commented on docs/src/clusters.md:
````diff
@@ -36,15 +36,15 @@ solana config set --url https://devnet.solana.com
 
 ```bash
 $ solana-validator \
+    --entrypoint entrypoint.devnet.solana.com:8001 \
````
Reordered for logical use order: entrypoint (contact to the cluster) => [trusted] validator (fetch genesis/snapshot) => expected-... (let's assert expected things finally)

ryoqun commented on docs/src/clusters.md:
```diff
@@ -74,20 +74,21 @@ solana config set --url https://testnet.solana.com
 
 ##### Example `solana-validator` command-line
 
+[comment]: <> (UPDATE ci/live-cluster-sanity.sh TOO!!)
```
TIL: https://stackoverflow.com/a/20885980

ryoqun replied:
I noticed docusaus can't handle this correctly, this is the reason of failing travis build.

ryoqun replied:
this is now fixed.

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+    -d '{"jsonrpc":"2.0","id":1, "method":"validatorExit"}' \
+    http://localhost:18899
+
+  (sleep 3 && kill "$tail_pid") &
```
this trick realizes `set +e`-less elegant `wait`.

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+  ./net/ssh.sh "$instance_ip" mkdir cluster-sanity
+
+  validator_log="$cluster_label-validator.log"
+  ./net/ssh.sh "$instance_ip" -Llocalhost:18899:localhost:18899 ./solana-validator \
```
combined with `--private-rpc` and `--rpc-bind-address`, the exposure to the public internet is minimized.

ryoqun replied:
no longer needed

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+    show_log
+  done
+
+  echo "--- Monitoring validator $cluster_label"
```
should I also add the `catchup` phase?

ryoqun replied:
let's skip this for now. This will increase the test time.

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+    --trusted-validator 9QxCLckBiJc783jnMvXZubK4wH86Eqqvashtrwvcsgkv \
+    --expected-genesis-hash 4uhcVJyU9pJkvQyS88uRDiswHXSCkY3zQawwpjk2NsNY \
+    # for your pain-less copy-paste
+
```
I wonder it's nice to have to upload fetched snapshots to the buildkite as artifacts for reproducible testing if anything odd happens.

ryoqun replied:
done by uploading snapshots only if the build failed.

ryoqun commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+
+  (sleep 3 && kill "$tail_pid") &
+  kill_pid=$!
+  wait "$ssh_pid" "$tail_pid" "$kill_pid"
```
guard with `timeout N `

ryoqun replied:
well it turned out this is rather complicated. hint: `wait` must be shell builtin but `timeout` is just a normal command. let's skip this.

mvines commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+source ci/_
+source ci/rust-version.sh stable
+
+escaped_branch=$(echo "$BUILDKITE_BRANCH" | tr -c "[:alnum:]" - | sed -r "s#(^-*|-*head-*|-*$)##g")
```
If `BUILDKITE_BRANCH` is empty (like if `ci/live-cluster-sanity.sh` is run locally), set `escaped_branch` to `$(whoami)` perhaps?

ryoqun replied:
done

mvines commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+# ensure to delete leftover cluster
+./net/gce.sh delete -p "$instance_prefix" || true
+# only bootstrap, no normal validator
+./net/gce.sh create -p "$instance_prefix" -n 0
```
Let's ensure the instances are shut down promptly if something goes wrong:
```suggestion
./net/gce.sh create -p "$instance_prefix" -n 0 --self-destruct-hours 1
```

ryoqun replied:
thanks for tipping this nice option. I didn't know it.

mvines commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+
+_ cargo +"$rust_stable" build --bins --release
+_ ./net/scp.sh ./target/release/solana-validator "$instance_ip:."
+echo 500000 | ./net/ssh.sh "$instance_ip" sudo tee /proc/sys/vm/max_map_count > /dev/null
```
Instead of this, let's copy `solana-sys-tuner` in so it can set max_map_count and we verify that code path too

ryoqun replied:
done

mvines commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+  fi
+}
+
+test_with_live_cluster() {
```
I find all the `./net/ssh.sh` calls in this function a little hard to follow.  I think it would be more clear if we copied a single shell script representing most/all of this function into the instance and just ran it.

ryoqun replied:
this is done!

mvines commented on docs/src/clusters.md:
````diff
@@ -36,15 +36,15 @@ solana config set --url https://devnet.solana.com
 
 ```bash
````
The doc/ and `bank.rs` changes in here look just fine, why don't you just land those as a separate PR while we work through the `ci/` files in this PR

ryoqun replied:
There is no strong reason to create separate PRs. I just thought it's not worth to be its own prs. `bank.rs` changes are needed to trigger `live-cluster` tests. (yeah, I could improve the `ci/buildkite-pipeline.sh`). And `docs` chagnes somewhat mentions this pr about the update notice. So separating them introduces a bit of work.

t-nelson commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,142 @@
+source ci/rust-version.sh stable
+
+escaped_branch=$(echo "$BUILDKITE_BRANCH" | tr -c "[:alnum:]" - | sed -r "s#(^-*|-*head-*|-*$)##g")
+instance_prefix="testnet-live-sanity-$escaped_branch"
```
I think metrics will complain about this since there won't be a database named `$instance_prefix`. (I hit similar trying to get cute with the rolling upgrades instance names)

ryoqun replied:
Fortunately, this pr doesn't use much of `net/*.sh`s. This isn't affected. Anyway, I've specifically setup a metric database for this job.

t-nelson commented on ci/live-cluster-sanity.sh:
```diff
@@ -0,0 +1,144 @@
+instance_ip=$(./net/gce.sh info | grep bootstrap-validator | awk '{print $3}')
+
+on_trap() {
+  if [[ -z $instance_deleted ]]; then
```
I think it's safe to just try to delete here

ryoqun commented on ci/buildkite-pipeline.sh:
```diff
@@ -169,6 +172,12 @@ all_test_steps() {
     artifact_paths: "log-*.txt"
     agents:
       - "queue=cuda"
+  - command: "ci/live-cluster-sanity.sh"
```
Yeah, I think this worth on every PR. These are some reasons:

- nightly is a bit too infrequent in my opinion;
  - According to [the insights](https://github.com/solana-labs/solana/pulse), it seems that we're merging 20 PRs per business day (100 per weak/500 per month). Then, assume roughly half of it is rust (validator) related (quick guess from https://buildkite.com/solana-labs/solana/builds?branch=master&page=2). Under that numbers in mind, bisecting regressions will take about 3 steps (2 ** 3 =~ 10) in average with nightly. This is tedious in my opinion; bisecting is very effective for the very-wide window, it's not so much effective in small window. 
    - I can tolerate with hourly, but then why not every-pr? ;)
  - This doesn't make the whole CI longer from the PR author's perspective (`local-cluster` is the longest at this pipeline phase...)
- it's less ideal compared to unit-tests, but this test could serve as a smoke test around process startup, whose tests are currently particularly weak.
- Running every PR could work as a last minute sanity check in the case of hotfix.
- `live-cluster` occupies `queue=gce-deploy` which isn't so crowded compared to the `queue=default`.

t-nelson replied:
> `live-cluster` occupies `queue=gce-deploy` which isn't so crowded compared to the `queue=default`.

I believe there are only one or two agents running `gce-deploy` ATM.  So we'll want to bump that up first.  It should just be a matter of ensuring the gcloud CLI tools are installed and pointed at the correct project, then adding a systemd service for the new agent

ryoqun commented:
@mvines @t-nelson I think this is good for another review in code-wise. The only remaining job is to increase build-agent for `gce-deploy`.

t-nelson reviewed (approved):
LGTM, now!  Rolling out more BK agents is currently blocked on https://github.com/solana-labs/solana/pull/12527, though

mvines reviewed (approved):
I removed the v1.2 label, I don't think we need that backport since we're about to end-of-life v1.2

ryoqun commented:
just fyi: this is blocked by #12527

ryoqun commented:
rebased and reverted some code commented-out for faster ci testing.

FYI: my intel agent (@carllin) tipped about another possible incompatibility of tip of master with mainnet-beta. let's test it. :)

ryoqun commented:
@carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:

https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69

```
thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
```

it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.

ryoqun commented:
> @carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:
> 
> https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69
> 
> ```
> thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
> ```
> 
> it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.

@jackcmay Is there something to ring in your mind. it seems that `solana-ledger-tool` still `dlopen()`s the `libsolana_bpf_loader_program.so`, which no longer build. See the `mainnet-beta/validator.log` artifact.

jackcmay commented:
@ryoqun Is this a single node test?  I don't see any dependency on bpf_loader in the ledger tool.  If could be the bank doesn't have the bpf_loader builtin?

ryoqun commented:
@jackcmay Yeah, this is a single node test, which connects to the actual mainnet-beta and testnet. I think there should be at least dependency on `bpf_loader` in the ledger tool. When `ledger-tool` replays the snapshot and following blocks with its `verify` subcommand and it encounters any program transactions in those blocks, it must execute those instructions. The referenced build test is failing at `ledger-tool verify`. Oddly enough, it seems `solana-validator` can process same set of transactions without problem (the test runs `solana-validator` before `solana-ledger-tool verify`).

ryoqun commented:
> > @carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:
> > https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69
> > ```
> > thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
> > ```
> > 
> > 
> > it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.
> 
> @jackcmay Is there something to ring in your mind. it seems that `solana-ledger-tool` still `dlopen()`s the `libsolana_bpf_loader_program.so`, which no longer build. See the `mainnet-beta/validator.log` artifact.

@jackcmay This will be fixed by #12759

jackcmay commented:
Awesome, thanks @ryoqun @t-nelson !

ryoqun commented:
> LGTM, now! Rolling out more BK agents is currently blocked on #12527, though

@t-nelson Regarding this, you and I seem to too busy to fix #12527... :) How about merging this by **making `live-cluster` run only on release branches (master, v1.4 and v1.3), not on pull request branches**? So just one or two line `skip if` in `live-cluster-sanity.sh`. So that the CI pressure should be low and we can tolerate with two `gce-deploy` agents as-is without fixing #12527 first. Or #12527 must be fixed regardless of the number of `gce-deploy`?

I think it's cost-efficient and good compromise (or sweet spot) between every-pr and nightly... ;) At least, we can avoid releasing blatant incompatible releases. (This occurred thrice recently already #12176, #12791 and #12759 and could be avoided with this pr). I want to boost our team's productivity by freeing them from manually running validators for sanity checking. And I want to free myself from digging those regressions as well. :)

t-nelson commented:
Any idea what the average run time is?

ryoqun commented:
> Any idea what the average run time is?

Around ~15 min real time. Because this is parallel to `local-cluster` (which takes over 20min), this shouldn't affect overall build time, assuming there is no `gce-deploy` contention. Under somewhat emergency, we can kill `master` CI and make `v1.3` and `v1.4` take the agents respectively from the 2 `gce-deploy` agents.

t-nelson commented:
I fear we'll get backed up on https://buildkite.com/solana-labs/system-performance-tests/settings/schedules.  That 8hr test is going to hurt over there in Asia :wink:

ryoqun commented:
> I fear we'll get backed up on https://buildkite.com/solana-labs/system-performance-tests/settings/schedules. That 8hr test is going to hurt over there in Asia wink

got it. so we're back to square one. let's the correct thing done. ;)
